josh-json/src/files.js

```diff
@@ -0,0 +1,124 @@
+const atomic = require('atomically');
+const path = require('path');
+const { serializeData, parseData } = require('./utils');
+const fs = require('fs');
+const { isEqual } = require('lodash');
+class FileManager {
+  constructor(dir, opts) {
+    this.options = opts || {};
+    this.dir = path.resolve(process.cwd(), dir);
+    try {
+      fs.mkdirSync(dir);
+    } catch {
+      // don't need to do anything
+    }
+  }
+  async getIndex(key) {
+    let index = await this.getFile('index.json');
+    if (
+      !index.files ||
+      !index.files.find((x) => x.keys.length < (this.options.maxLength || 100))
+    ) {
+      index = await this.addFile(index);
+    }
+    const found = index.files.find((x) => x.keys.includes(key));
+    if (key && found) {
+      return found;
+    } else {
+      return index;
+    }
+  }
+  async setIndex(data) {
+    return await this.setFile('index.json', data);
+  }
+  async setFile(file, data) {
+    await atomic.writeFile(path.resolve(this.dir, file), serializeData(data));
+    return data;
+  }
+  async getFile(file) {
+    let data = {};
+    try {
+      data = parseData(
+        await atomic.readFile(path.resolve(this.dir, file), {
+          encoding: 'utf8',
+        }),
+      );
+    } catch (err) {
+      await atomic.writeFile(path.resolve(this.dir, file), '{}');
+    }
+    return data;
+  }
+  async getData(key) {
+    const index = await this.getIndex(key);
+    if (key) {
+      if (index && index.location) {
+        return (await this.getFile(index.location))[key];
+      } else {
+        return null;
+      }
+    }
+    let all = {};
+    for (const file of index.files) {
+      all = { ...all, ...await this.getFile(file.location) };
+    }
+    return all;
+  }
+  async addFile(index) {
+    if (!index) index = await this.getIndex();
+    if (!index.files) {
+      index.files = [];
+    }
+    const location = path.resolve(this.dir, `josh_${index.files.length}.json`);
```

Comment:
Oh I see what you're doing for indexes now. I feel like having the name of the file itself be the key (with a "clean" function as I have with sqlite keys) might be faster and better?

Reply:
I thought about that but each file stores up to 100 keys, (options.maxLength can change this)